### DISCUSSION TITLE
UZF-GWET term on volume budget

### DISCUSSION THREAD

**Discussion by martclanor** [MODFLOW-ORG/modflow6#1904] [Q&A]
Hello all. Perhaps a long shot but I was wondering if anyone can help me make sense how come the UZF-GWET is an inflow instead of an outflow. What does that mean in practice? What could have possibly caused this?

This is the Volume Budget section of my LST file (I'm also attaching the file itself, in case it can help [test_uzf.zip](https://github.com/user-attachments/files/15990871/test_uzf.zip)):

```
  UZF BUDGET FOR ENTIRE MODEL AT END OF TIME STEP    1, STRESS PERIOD   1
  ------------------------------------------------------------------------------

     CUMULATIVE UZF      L**3       RATES FOR THIS TIME STEP      L**3/T
     ------------------                 ------------------------

           IN:                                      IN:
           ---                                      ---
                 GWF =           0.0000                   GWF =           0.0000
        INFILTRATION =    41852715.4406          INFILTRATION =      114664.9738
             REJ-INF =           0.0000               REJ-INF =           0.0000
                UZET =           0.0000                  UZET =           0.0000
             STORAGE =           0.0000               STORAGE =           0.0000
           AUXILIARY =           0.0000             AUXILIARY =           0.0000

            TOTAL IN =    41852715.4406              TOTAL IN =      114664.9738

          OUT:                                     OUT:
          ----                                     ----
                 GWF =    43640277.6320                   GWF =      119562.4045
        INFILTRATION =           0.0000          INFILTRATION =           0.0000
             REJ-INF =     2047538.9869               REJ-INF =        5609.6959
                UZET =           0.0000                  UZET =           0.0000
             STORAGE =           0.0000               STORAGE =           0.0000
           AUXILIARY =           0.0000             AUXILIARY =           0.0000

           TOTAL OUT =    45687816.6189             TOTAL OUT =      125172.1003

            IN - OUT =    -3835101.1783              IN - OUT =      -10507.1265

 PERCENT DISCREPANCY =          -8.76     PERCENT DISCREPANCY =          -8.76



  VOLUME BUDGET FOR ENTIRE MODEL AT END OF TIME STEP    1, STRESS PERIOD   1
  ---------------------------------------------------------------------------------------------------

     CUMULATIVE VOLUME      L**3       RATES FOR THIS TIME STEP      L**3/T          PACKAGE NAME    
     ------------------                 ------------------------                     ----------------

           IN:                                      IN:
           ---                                      ---
              STO-SS =           0.0000                STO-SS =           0.0000     STORAGE                          
              STO-SY =           0.0000                STO-SY =           0.0000     STORAGE                          
                 WEL =           0.0000                   WEL =           0.0000     WEL_KLEINE_ONT                   
                 WEL =           0.0000                   WEL =           0.0000     WEL_GROTE_ONT                    
                 WEL =           0.0000                   WEL =           0.0000     WEL_DWATER_ONT                   
                 DRN =           0.0000                   DRN =           0.0000     DRN                              
                 DRN =           0.0000                   DRN =           0.0000     DRN_LANDBOUW                     
                 DRN =           0.0000                   DRN =           0.0000     DRN_MAAIVELD                     
                 RIV =   914174904.9653                   RIV =     2504588.7807     RIV_VL                           
                 RIV =   767810433.9801                   RIV =     2103590.2301     RIV_NL                           
                 CHD =     6201944.7010                   CHD =       16991.6293     CHD                              
                 LAK =           0.0000                   LAK =           0.0000     LAK                              
           UZF-GWRCH =    43640277.6320             UZF-GWRCH =      119562.4045     UZF                              
            UZF-GWET =   112491297.3466              UZF-GWET =      308195.3352     UZF                              

            TOTAL IN =  1844318858.6250              TOTAL IN =     5052928.3798

          OUT:                                     OUT:
          ----                                     ----
              STO-SS =           0.0000                STO-SS =           0.0000     STORAGE                          
              STO-SY =           0.0000                STO-SY =           0.0000     STORAGE                          
                 WEL =      549432.8184                   WEL =        1505.2954     WEL_KLEINE_ONT                   
                 WEL =     8619200.0000                   WEL =       23614.2466     WEL_GROTE_ONT                    
                 WEL =    14852834.5884                   WEL =       40692.6975     WEL_DWATER_ONT                   
                 DRN =    29817881.2571                   DRN =       81692.8254     DRN                              
                 DRN =  1457357609.6451                   DRN =     3992760.5744     DRN_LANDBOUW                     
                 DRN =     7578019.2530                   DRN =       20761.6966     DRN_MAAIVELD                     
                 RIV =    56810703.1939                   RIV =      155645.7622     RIV_VL                           
                 RIV =   259600781.9585                   RIV =      711235.0191     RIV_NL                           
                 CHD =     5914667.4119                   CHD =       16204.5683     CHD                              
                 LAK =           0.0000                   LAK =           0.0000     LAK                              
           UZF-GWRCH =           0.0000             UZF-GWRCH =           0.0000     UZF                              
            UZF-GWET =           0.0000              UZF-GWET =           0.0000     UZF                              


           TOTAL OUT =  1841101130.1262             TOTAL OUT =     5044112.6853

            IN - OUT =     3217728.4988              IN - OUT =        8815.6945

 PERCENT DISCREPANCY =           0.17     PERCENT DISCREPANCY =           0.17


```

And also if

`UZF BUDGET: OUT: GWF` = `VOLUME BUDGET: IN: UZF-GWRCH` = `119562.4045`

then where is:

`VOLUME BUDGET: IN: UZF-GWET` = `308195.3352`

in `UZF BUDGET: OUT`?

**Comment by christianlangevin** [comment]
Not sure what is going on here, but i see in your list file that a bunch of cells went dry.  In theory, that might be okay if rewetting is on and the answer is good, but in practice a bunch of dry cells is not ideal.  I'd suggest you turn on NEWTON and try solving this problem differently.  The positive UZF-GWET term is strange, and something doesn't seem right so if that persists we should have a look.

**Reply by martclanor** [reply]
Thanks a lot already for answering. For the moment, I couldn't come up with a converging solution with the NEWTON turned on for this steady-state model. However, I also have a transient version of this model which is already run with the NEWTON turned on and where this positive UZF-GWET issue persists for all stress periods. Can you also maybe have a look on this?
[test_uzf_unsteady.zip](https://github.com/user-attachments/files/16012741/test_uzf_unsteady.zip)

**Reply by martclanor** [reply]
I am not sure about how to attach the UZF file, as the one that I have for this model is 2.7GB.

A bit tangent to the topic here but I was unable to implement UZF timeseries in this model without significantly increasing run times. This might be related to this issue: https://github.com/MODFLOW-USGS/modflow6/issues/1338, but I've been holding off on reopening the issue until I can double check this.

**Comment by emorway-usgs** [comment]
If, after making the previously suggested changes, you are still seeing such a large percent discrepancy in the UZF budget, it might be worth adding the `PRINT_BUDGET` keyword to the `OPTIONS` block and focusing attention on specific UZF objects with their own bad budgets.  I've not encountered `UZF-GWET` showing up on the inflow side of the budget before.

**Reply by martclanor** [reply]
Thanks for this suggestion. I assume here that you meant `PRINT_FLOWS` in the `OPTIONS` block of `UZF`? Using this setting, I have this output on the list file now (see complete extracted table here: [uzf_summary_of_flows.csv](https://github.com/user-attachments/files/16013899/uzf_summary_of_flows.csv)).

```
 UZF PACKAGE - SUMMARY OF FLOWS FOR EACH CONTROL VOLUME   PERIOD      1   STEP        1
 ----------------------------------------------------------------------------------------------------------------------------------------------------
                                                                                                                                           PERCENT   
   NUMBER   CELLID                  INFLOW      OUTFLOW        GWF      INFILTRATION   REJ INF        UZET       STORAGE      IN - OUT    DIFFERENCE 
 ----------------------------------------------------------------------------------------------------------------------------------------------------
     1      (3,1)                 0.0000       0.0000      -55.052       55.052       0.0000       0.0000       0.0000       0.0000       0.0000     
     2      (3,2)                 0.0000       0.0000      -55.052       55.052       0.0000       0.0000       0.0000       0.0000       0.0000     
     3      (3,3)                 0.0000       0.0000      -49.132       49.132       0.0000       0.0000       0.0000       0.0000       0.0000     
     4      (3,4)                 0.0000       0.0000      -55.052       55.052       0.0000       0.0000       0.0000       0.0000       0.0000     
     5      (3,5)                 0.0000       0.0000      -49.132       49.132       0.0000       0.0000       0.0000       0.0000       0.0000     
     6      (3,6)                 0.0000       0.0000      -55.052       55.052       0.0000       0.0000       0.0000       0.0000       0.0000     
     7      (3,7)                 0.0000       0.0000      -57.419       57.419       0.0000       0.0000       0.0000       7.10543E-15  1.23746E-14
     8      (3,8)                 0.0000       0.0000      -55.052       55.052       0.0000       0.0000       0.0000       0.0000       0.0000     
     9      (3,9)                 0.0000       0.0000      -52.684       52.684       0.0000       0.0000       0.0000      -7.10543E-15 -1.34869E-14
     10     (3,10)                0.0000       0.0000      -52.684       52.684       0.0000       0.0000       0.0000      -7.10543E-15 -1.34869E-14
     11     (3,11)                0.0000       0.0000      -52.684       52.684       0.0000       0.0000       0.0000      -7.10543E-15 -1.34869E-14
     12     (3,12)                0.0000       0.0000      -55.052       55.052       0.0000       0.0000       0.0000       0.0000       0.0000     
     13     (3,13)                0.0000       0.0000      -17.759       17.759       0.0000       0.0000       0.0000       0.0000       0.0000     
     14     (3,14)                0.0000       0.0000      -52.684       52.684       0.0000       0.0000       0.0000      -7.10543E-15 -1.34869E-14
     15     (3,15)                0.0000       0.0000      -55.052       55.052       0.0000       0.0000       0.0000       0.0000       0.0000     
     16     (1,16)                0.0000       0.0000      -55.052       0.0000       0.0000       0.0000       0.0000      -55.052      -200.00     
     17     (3,16)                0.0000       0.0000       0.0000       0.0000       0.0000       0.0000       0.0000       0.0000       0.0000     
```

Please note that this output is not anymore exactly the same model as in the original question (I apparently posted an old version, apologies) but the same problem persists.

Tried looking around and here are some notes from the attached csv:

1. In 63% of the cells, `GWF = -1*INFILTRATION`
2. In 94% of the cells with REJ-INF, `INFILTRATION = -1*REJ-INF`
3. `INFLOW`, `OUTFLOW`, `UZET` and `STORAGE` columns are just zero for all cells
4. The `PERCENT DIFFERENCE` column gives `-200` whenever there's a significant difference in `IN-OUT` column
5. For some cells e.g. `3789`, `4818`, etc., `GWF = -59.195` then all the other water balance terms are equal to zero
6. `-1*min(GWF)` = `max(INFILTRATION)` = `-1*min(REJ INF)` = `-1*min(IN-OUT)` = `59.195`

Any of these perhaps point to the issue?

**Reply by emorway-usgs** [reply]
Perhaps attach a copy of the UZF input file that includes everything through the first stress period and delete the rest.  It seems all of your  specified infiltration is passing through the UZF object and going straight to recharge.  With regard to the cell that has a 200% discrepancy, that water may be draining from the UZF object which would mean that there should be some inflow from storage.  Perhaps double check that the storage related parameters are as you intended.

**Reply by martclanor** [reply]
Here is a copy of the related UZF file: [no_newt_40.zip](https://github.com/user-attachments/files/16029895/no_newt_40.zip)

Here are the storage-related parameters that were used:
```python
# Block: PACKAGEDATA
thtr = 0.03  # residual wc
thts = 0.4  # saturated wc
thti = 0.15  # initial wc

# Block: PERIOD
extwc = 0.03  # ET extinction wc
```

**Reply by martclanor** [reply]
I will follow this up with the truncated UZF input file and list file for an unsteady model run (with Newton) as soon as it finishes.

**Comment by martclanor** [comment]
An update on this which might be interesting:

I did two model runs where the only difference is that one is `square_gwet = True` while the other is `linear_gwet =True`

## SQUARE GWET

- list and uzf files: [no_newt_40.zip](https://github.com/user-attachments/files/16030052/no_newt_40.zip)
<details>
<summary>volume budget</summary>

  UZF BUDGET FOR ENTIRE MODEL AT END OF TIME STEP    1, STRESS PERIOD   1
  ------------------------------------------------------------------------------

     CUMULATIVE UZF      L**3       RATES FOR THIS TIME STEP      L**3/T
     ------------------                 ------------------------

           IN:                                      IN:
           ---                                      ---
                 GWF =           0.0000                   GWF =           0.0000
        INFILTRATION =   120391888.5863          INFILTRATION =      329840.7906
             REJ-INF =           0.0000               REJ-INF =           0.0000
                UZET =           0.0000                  UZET =           0.0000
             STORAGE =           0.0000               STORAGE =           0.0000

            TOTAL IN =   120391888.5863              TOTAL IN =      329840.7906

          OUT:                                     OUT:
          ----                                     ----
                 GWF =   163094101.8070                   GWF =      446833.1556
        INFILTRATION =           0.0000          INFILTRATION =           0.0000
             REJ-INF =     8502607.5959               REJ-INF =       23294.8153
                UZET =           0.0000                  UZET =           0.0000
             STORAGE =           0.0000               STORAGE =           0.0000

           TOTAL OUT =   171596709.4029             TOTAL OUT =      470127.9710

            IN - OUT =   -51204820.8165              IN - OUT =     -140287.1803

 PERCENT DISCREPANCY =         -35.07     PERCENT DISCREPANCY =         -35.07



  VOLUME BUDGET FOR ENTIRE MODEL AT END OF TIME STEP    1, STRESS PERIOD   1
  ---------------------------------------------------------------------------------------------------

     CUMULATIVE VOLUME      L**3       RATES FOR THIS TIME STEP      L**3/T          PACKAGE NAME    
     ------------------                 ------------------------                     ----------------

           IN:                                      IN:
           ---                                      ---
              STO-SS =           0.0000                STO-SS =           0.0000     STORAGE                          
              STO-SY =           0.0000                STO-SY =           0.0000     STORAGE                          
                 WEL =           0.0000                   WEL =           0.0000     WEL_KLEINE_ONT                   
                 WEL =           0.0000                   WEL =           0.0000     WEL_GROTE_ONT                    
                 WEL =           0.0000                   WEL =           0.0000     WEL_DWATER_ONT                   
                 DRN =           0.0000                   DRN =           0.0000     DRN                              
                 DRN =           0.0000                   DRN =           0.0000     DRN_LANDBOUW                     
                 DRN =           0.0000                   DRN =           0.0000     DRN_MAAIVELD                     
                 RIV =     4620917.0808                   RIV =       12660.0468     RIV_VL                           
                 RIV =     2555036.7746                   RIV =        7000.1008     RIV_NL                           
                 CHD =     2798712.3479                   CHD =        7667.7051     CHD                              
                 LAK =           0.0000                   LAK =           0.0000     LAK                              
           UZF-GWRCH =   163094101.8070             UZF-GWRCH =      446833.1556     UZF                              
            UZF-GWET =   231379426.4097              UZF-GWET =      633916.2367     UZF                              

            TOTAL IN =   404448194.4201              TOTAL IN =     1108077.2450

          OUT:                                     OUT:
          ----                                     ----
              STO-SS =           0.0000                STO-SS =           0.0000     STORAGE                          
              STO-SY =           0.0000                STO-SY =           0.0000     STORAGE                          
                 WEL =      549432.8184                   WEL =        1505.2954     WEL_KLEINE_ONT                   
                 WEL =     8619200.0000                   WEL =       23614.2466     WEL_GROTE_ONT                    
                 WEL =    14852834.5884                   WEL =       40692.6975     WEL_DWATER_ONT                   
                 DRN =    62394251.3754                   DRN =      170943.1545     DRN                              
                 DRN =   220314376.2229                   DRN =      603601.0307     DRN_LANDBOUW                     
                 DRN =    24096106.5894                   DRN =       66016.7304     DRN_MAAIVELD                     
                 RIV =     9823336.0190                   RIV =       26913.2494     RIV_VL                           
                 RIV =     5350439.6440                   RIV =       14658.7388     RIV_NL                           
                 CHD =     8095284.6808                   CHD =       22178.8621     CHD                              
                 LAK =           0.0000                   LAK =           0.0000     LAK                              
           UZF-GWRCH =           0.0000             UZF-GWRCH =           0.0000     UZF                              
            UZF-GWET =           0.0000              UZF-GWET =           0.0000     UZF                              

           TOTAL OUT =   354095261.9384             TOTAL OUT =      970124.0053

            IN - OUT =    50352932.4816              IN - OUT =      137953.2397

 PERCENT DISCREPANCY =          13.28     PERCENT DISCREPANCY =          13.28


</details>


## LINEAR GWET

- list and uzf files: [no_newt_40_line.zip](https://github.com/user-attachments/files/16030078/no_newt_40_line.zip)
<details>
<summary>volume budget</summary>

  UZF BUDGET FOR ENTIRE MODEL AT END OF TIME STEP    1, STRESS PERIOD   1
  ------------------------------------------------------------------------------

     CUMULATIVE UZF      L**3       RATES FOR THIS TIME STEP      L**3/T
     ------------------                 ------------------------

           IN:                                      IN:
           ---                                      ---
                 GWF =           0.0000                   GWF =           0.0000
        INFILTRATION =    32537901.8021          INFILTRATION =       89144.9364
             REJ-INF =           0.0000               REJ-INF =           0.0000
                UZET =           0.0000                  UZET =           0.0000
             STORAGE =           0.0000               STORAGE =           0.0000

            TOTAL IN =    32537901.8021              TOTAL IN =       89144.9364

          OUT:                                     OUT:
          ----                                     ----
                 GWF =   171328553.0737                   GWF =      469393.2961
        INFILTRATION =           0.0000          INFILTRATION =           0.0000
             REJ-INF =      268156.3291               REJ-INF =         734.6749
                UZET =           0.0000                  UZET =           0.0000
             STORAGE =           0.0000               STORAGE =           0.0000

           TOTAL OUT =   171596709.4029             TOTAL OUT =      470127.9710

            IN - OUT =  -139058807.6007              IN - OUT =     -380983.0345

 PERCENT DISCREPANCY =        -136.24     PERCENT DISCREPANCY =        -136.24



  VOLUME BUDGET FOR ENTIRE MODEL AT END OF TIME STEP    1, STRESS PERIOD   1
  ---------------------------------------------------------------------------------------------------

     CUMULATIVE VOLUME      L**3       RATES FOR THIS TIME STEP      L**3/T          PACKAGE NAME    
     ------------------                 ------------------------                     ----------------

           IN:                                      IN:
           ---                                      ---
              STO-SS =           0.0000                STO-SS =           0.0000     STORAGE                          
              STO-SY =           0.0000                STO-SY =           0.0000     STORAGE                          
                 WEL =           0.0000                   WEL =           0.0000     WEL_KLEINE_ONT                   
                 WEL =           0.0000                   WEL =           0.0000     WEL_GROTE_ONT                    
                 WEL =           0.0000                   WEL =           0.0000     WEL_DWATER_ONT                   
                 DRN =           0.0000                   DRN =           0.0000     DRN                              
                 DRN =           0.0000                   DRN =           0.0000     DRN_LANDBOUW                     
                 DRN =           0.0000                   DRN =           0.0000     DRN_MAAIVELD                     
                 RIV =    10758795.5766                   RIV =       29476.1523     RIV_VL                           
                 RIV =     3619505.7313                   RIV =        9916.4541     RIV_NL                           
                 CHD =    24118336.4562                   CHD =       66077.6341     CHD                              
                 LAK =           0.0000                   LAK =           0.0000     LAK                              
           UZF-GWRCH =   171328553.0737             UZF-GWRCH =      469393.2961     UZF                              
            UZF-GWET =           0.0000              UZF-GWET =           0.0000     UZF                              

            TOTAL IN =   209825190.8379              TOTAL IN =      574863.5365

          OUT:                                     OUT:
          ----                                     ----
              STO-SS =           0.0000                STO-SS =           0.0000     STORAGE                          
              STO-SY =           0.0000                STO-SY =           0.0000     STORAGE                          
                 WEL =      549432.8184                   WEL =        1505.2954     WEL_KLEINE_ONT                   
                 WEL =     8619200.0000                   WEL =       23614.2466     WEL_GROTE_ONT                    
                 WEL =    14852834.5884                   WEL =       40692.6975     WEL_DWATER_ONT                   
                 DRN =     5999465.1658                   DRN =       16436.8909     DRN                              
                 DRN =    17731795.9743                   DRN =       48580.2629     DRN_LANDBOUW                     
                 DRN =     1067026.5840                   DRN =        2923.3605     DRN_MAAIVELD                     
                 RIV =     1882142.6920                   RIV =        5156.5553     RIV_VL                           
                 RIV =     1078157.6846                   RIV =        2953.8567     RIV_NL                           
                 CHD =      382100.6870                   CHD =        1046.8512     CHD                              
                 LAK =           0.0000                   LAK =           0.0000     LAK                              
           UZF-GWRCH =           0.0000             UZF-GWRCH =           0.0000     UZF                              
            UZF-GWET =    19171286.2806              UZF-GWET =       52524.0720     UZF                              

           TOTAL OUT =    71333442.4752             TOTAL OUT =      195434.0890

            IN - OUT =   138491748.3627              IN - OUT =      379429.4476

 PERCENT DISCREPANCY =          98.52     PERCENT DISCREPANCY =          98.52

</details>

---

With the `LINEAR GWET`, `UZF-GWET` now becomes an outflow term, but `PERCENT DISCREPANCY` is increased. Tbh, I'm not sure if this is a step in the right direction, but perhaps this gives a hint about the root cause?

**Reply by christianlangevin** [reply]
In the table that you show above `UZF PACKAGE - SUMMARY OF FLOWS FOR EACH CONTROL VOLUME   PERIOD      1   STEP        1` do you see anything strange happening for UZF cell 16, which is connected to cellid (1,16)?  Is cellid (1,16) dry?  What is the water table elevation in that cell?  Are there multiple model layers?  Is there an active cell beneath (1,16)?  I think there is a clue here somewhere.

**Reply by martclanor** [reply]
> Is cellid (1,16) dry? What is the water table elevation in that cell?

Yes, it's dry.
```python
# Using flopy
heads = gwf.output.head().get_alldata()[0, :, 0, :]
heads[0, 15]  # head at cellid (1, 16) = -1e+30
```
---
> Are there multiple model layers?

Yes, there are 16 model layers

---

> Is there an active cell beneath (1,16)?

Yes there are.
```python
idomain[:, 15]
# array([ 1, -1,  1, -1, -1,  1,  1,  1,  1,  1,  1,  1,  1, -1,  1,  1])

heads[:, 15] 
# array([-1.00000000e+30,  1.00000000e+30, -1.00000000e+30,  1.00000000e+30,
#         1.00000000e+30,  5.86177451e+00,  5.86177539e+00,  5.86088882e+00,
#         5.85999232e+00,  5.85992397e+00,  5.85991311e+00,  5.83819604e+00,
#         5.81637054e+00,  1.00000000e+30,  5.81627201e+00,  5.81611369e+00])

```
---

So in the first 6 layers on this vertical column (n, 16):

- layers 1 and 3 are UZF cells (UZF cell 16 and UZF cell 17) (cellid 1,16 and cellid 1,17)
- layers 2, 4 and 5 are inactive cells
- layers 6 has a head

**Reply by christianlangevin** [reply]
Okay, this is very helpful.  It's helpful to know that you are using IDOMAIN=-1 cells and that drying is involved.  My hunch is that you will need to use Newton for this problem.  I'm not sure what UZF is doing when the GWF cell is dry, but I'm guessing that is the root cause of the problem.

**Reply by martclanor** [reply]
Thanks a lot for this insight! I will look further in this direction and will post an update as soon as I have something.

**Reply by emorway-usgs** [reply]
@martclanor, Thanks for posting the UZF input data.  Building upon what @langevin-usgs pointed out (hopefully), a couple of minor things I'm noticing:
- Having looked at the input you posted, I'm a little confused by the budget you posted prior to that, but maybe that's because of what you stated, 

> "Please note that this output is not anymore exactly the same model as in the original question (I apparently posted an old version, apologies) but the same problem persists".  

Thought I had a look in the *.lst file you shared and am seeing the same issue: you specify a `FINF` value for UZF 16, but the `INFILTRATION` term in the budget you shared is 0?
- For UZF object 17, situated just below 16, all the budget terms are 0.0, though it should be receiving the drainage from UZF 16? 
- Something to consider is whether you want to add vertical connection to the next active cell below UZF 17 so that vertically draining water in the unsat zone continues being routed downward (ignore if this conflicts with your conceptual model) to the next cell in the stack (layer 6)?  
- For UZF objects vertically connected to an overlying UZF object (situated below), best to set `SURFDEP` equal to 0 since there are no surface undulations to be represented for this situation.  In other words, I'm not sure it makes sense to set `SURFDEP` = 1e-6 for UZF 17?  However, I don't think this will be causing any of the present issues.

**Reply by martclanor** [reply]
@emorway-usgs, thanks for checking the files. I just run a transient version of the model with Newton (as @langevin-usgs  suggested). Some answers to your questions and thoughts:

---

> you specify a FINF value for UZF 16, but the INFILTRATION term in the budget you shared is 0?

Yes, this is indeed the case. In this new model run, now I can see that the calculated head in (1,16) is actually above the top elevation of the cell (which should explain why the `INFILTRATION` is all rejected). This case of the head being higher than the surface elevation is apparent in 79% of the top model cells.

Btw, in the model, there is also an implementation of drain with high conductivity on all the top surface elevation. But I'm not sure how that plays here because the head in UZF 16 (1,16) is higher than the drain.

---

> For UZF object 17, situated just below 16, all the budget terms are 0.0, though it should be receiving the drainage from UZF 16?

In this new model run, I noticed that UZF 17 actually has an even higher head than the UZF cell above it, UZF 16 (thus, also higher than the top surface drain). This is perhaps already dubious in and of itself, but what confuses me more is why doesn't the water flow upward from UZF 17 to UZF 16 to the drain. Is that not possible anymore if the cells is assigned as UZF?

---

>  add vertical connection to the next active cell below UZF 17 so that vertically draining water in the unsat zone continues being routed downward

I don't know for sure but I was assuming that this vertical connection is implicit for all UZF cells with `ivertcon = 0`. Is that not the case?

---

Not sure what this means but I thought I'll point this out: Going back to the issue of positive UZF-GWET term. I just noticed that the UZF-GWET for all UZF cells is greater than or equal to 0. None of the UZF cells has a negative UZF-GWET.

---

I think we're getting somewhere here but I can't put my finger on it just yet. Is the issue about the UZF cells being saturated? Curious on your thoughts about this. Again, thanks a lot already for helping.

**Reply by emorway-usgs** [reply]
A couple of responses:
> I don't know for sure but I was assuming that this vertical connection is implicit for all UZF cells with ivertcon = 0. Is that not the case?

There are no implicit connections in UZF.  Per the I/O guide: "_A value of 0 indicates the UZF cell is not connected to an
underlying UZF cell._"

> Not sure what this means but I thought I'll point this out: Going back to the issue of positive UZF-GWET term. I just noticed that the UZF-GWET for all UZF cells is greater than or equal to 0. None of the UZF cells has a negative UZF-GWET.

UZF-GWET will be reported as a positive value, the question is why it shows up in the inflow portion of the budget and not the outflow portion?  It may be due to the heads being above your specified ground surface elevation, I doubt this is a condition that was tested previously and is something I or someone on the development team should look more into.  If, in your particular application, the heads are not supposed to be above the ground elevation then it would be helpful to know if lowering the heads below the ground surface elevation helps fix the UZF-GWET budget issue.

> In this new model run, I noticed that UZF 17 actually has an even higher head than the UZF cell above it, UZF 16 (thus, also higher than the top surface drain). This is perhaps already dubious in and of itself, but what confuses me more is why doesn't the water flow upward from UZF 17 to UZF 16 to the drain. Is that not possible anymore if the cells is assigned as UZF?

If the cells are saturated, then UZF is playing no role in the groundwater flow calculations.  Instead, the GWF model is solving for groundwater flow among the active cells.

> Btw, in the model, there is also an implementation of drain with high conductivity on all the top surface elevation. But I'm not sure how that plays here because the head in UZF 16 (1,16) is higher than the drain.

I think it would help smooth the solution to lower the drains to some depth below ground surface elevation.  The idea of lowering the drain just a bit is to represent the land surface undulations within a cell.  There are some options available within the DRN package that may be helpful to you, please see the I/O guide for details.

**Reply by martclanor** [reply]
Thank you for these insights! Will work further with these in mind.

Meanwhile, another update:

I just noticed that SPD data in the model (including `finf` and `pet`) are being assigned to all UZF cells. Perhaps, adapting this to apply only to top UZF cells might do something especially for `PET` because based on the docs for UZF-perioddata-pet:

_If IVERTCON is greater than zero then residual potential evapotranspiration not satisfied in the UZF cell is applied to the underlying UZF and GWF cells._

**Reply by emorway-usgs** [reply]
> I just noticed that SPD data in the model (including finf and pet) are being assigned to all UZF cells. Perhaps, adapting this to apply only to top UZF cells might do something especially for PET because based on the docs for UZF-perioddata-pet:

Just to be clear, are you seeing `FINF` values being applied to cells that have a `landflag` value equal to 1 _and_ also to underlying cells?  Remember that in a steady state stress period the applied `FINF` rates will pass through each dry cell in a vertical stack of cells.  In the budget it will look like `FINF` is being applied to underlying cell, though it is just the steady state drainage from the overlying cell that happens to be equal to `FINF` in steady state stress period.

> If IVERTCON is greater than zero then residual potential evapotranspiration (`pET`) not satisfied in the UZF cell is applied to the underlying UZF and GWF cells.

If, in a vertical stack of UZF objects, the water table is in a lower layer that is within the specified `EXTDP` and not all of the `pET` is satisfied with UZET from the overlying cell(s) and the cell hosting the water table, then the residual `pET` may still be satisfied with ET from the water table (`GWET`).

**Reply by martclanor** [reply]
> Just to be clear, are you seeing FINF values being applied to cells that have a landflag value equal to 1 and also to underlying cells?

Short answer: no. `FINF` values are indeed only being applied to UZF cells with `landflag = 1`

Just did an experiment with two unsteady-state models:
- `MODEL 1`: `perioddata` (with `finf` and `pet` ) is assigned to all UZF cells
- `MODEL 2`: `perioddata` (with `finf` and `pet`) is assigned only to UZF cells which have `landflag = 1` in the `packagedata`

Results are not exactly the same, but just slightly different, and I don't see which is better. This path is maybe a dead-end but just a question about this, do I have to apply the perioddata for all UZF cells?

---

Just in case the answer to the previous question is yes, how do I determine the extinction depth, `extdp` for the underlying UZF cells? Do I use the same value as their corresponding overlying cells? Or do I have to offset this with the thickness of the overlying cell?

![image](https://github.com/MODFLOW-USGS/modflow6/assets/73061182/50746625-192a-451a-84cd-421ff81f1ae7)

So in this case, should it be:
- UZF cell 1: extdp 1
- UZF cell 2: extdp 1
- UZF cell 3: extdp 1

or
- UZF cell 1: extdp 1 
- UZF cell 2: extdp 1 - t1
- UZF cell 3: extdp 1 - t1 - t2

---

Maybe another possible clue: I found some cells whose head is lower than the cell's bottom. Is this probably related to the reversed direction of `GWET`? Note that this is for an unsteady-model with Newton.

---

Also is it worth going for the `linear_gwet` implementation instead of the `square_gwet`? Model version with `linear_gwet` results into `UZF-GWET` being in the outflow side (instead of inflow) although it still has some issues with convergence on some stress periods.

**Reply by emorway-usgs** [reply]
> Results are not exactly the same, but just slightly different, and I don't see which is better. This path is maybe a dead-end but just a question about this, do I have to apply the perioddata for all UZF cells?

Only apply `finf` and `pET` to the cells with `landflag=1`.  Specified boundary values like `finf` and `pET` are repeatedly used in successive stress periods if no overriding value is specified for the next stress period.  So for example, if the specified `pET` is 0.01 in stress period 1 and you need it to be 0.0 in stress period two, you will need to explicitly set that for the appropriate cells (or use an empty `PERIODDATA` block to set all values equal to 0.0).

> ...how do I determine the extinction depth, extdp for the underlying UZF cells?

You do not need to worry about this.  The `EXTDP` is applied starting from the cell with `landflag=1`.  For example, in a stack of UZF objects that are each 2 units thick, if you specify `EXTDP` to be 3 units, then the code will internally sort out whether the water table is intersected by the extinction depth.

>  ...I found some cells whose head is lower than the cell's bottom. Is this probably related to the reversed direction of GWET? 

No, if the head is below the cell bottom then the UZF package should be taking care of the various budget terms.  If you are able to recreate the budget issue with a small reproducible example I can run on my side, that will be the quickest way to get to the bottom of what might be going on.

> ...Also is it worth going for the `linear_gwet` implementation instead of the `square_gwet`.

Sorry but I cannot advise on how you should be modeling your particular setup.  However, if you are able to recreate the problem using a small example problem where the budget looks good with `linear_gwet` but does not with `square_gwet`, I'm keenly interested in running that model locally.

**Reply by martclanor** [reply]
Attached here is a downscaled version of the model with the same budget issue. 
[us_dummy_ic_s.zip](https://github.com/user-attachments/files/16097202/us_dummy_ic_s.zip)

Please let me know if this is workable enough or if I have to scale it down even more. Some notes about the model:

- It was challenging to have it converging, so I tried just using a more lenient outer_dvclose and this one is set to `outer_dvclose = 0.7`. Hope that's fine.
- The initial conditions used are based on the results of a steady-state run with the same model configuration.
- The model is unsteady with 4 spd, 1 ts each.
- It runs under 20-secs on my machine with mf 6.5.0.
- The resulting `UZF-GWET` term for all 4 spds are in the inflow side
- I can also confirm that upon switching from `square_gwet` to `linear_gwet`, the `UZF-GWET` term is moved to the outflow side.

Thanks a lot in advance!

**Comment by christianlangevin** [comment]
Hey @martclanor, you mention that you are seeing this behavior for both steady state and transient simulations.  UZF behaves very differently for these two cases.  For steady state simulations, UZF does not track waves and instead does a simple calculation as described in the UZF1 Documentation, shown in italics below.  

_Steady state can be specified for the initial stress period of a simulation when using the UZF1 Package. When steady state is specified for the initial stress period, the ground-water recharge rate computed in UZF1 is set equal to the infiltration rate and ET is not removed from the unsaturated zone. The water content is constant within each model cell during steady state. The specified ET rate is subtracted from the infiltration rate to calculate a steady-state recharge rate. ET can also occur from ground water when the water table is above the ET extinction depth. A uniform water content profile is computed for each active cell, which is used automatically as the initial conditions for a following transient stress period._

_Steady-state water content profiles may be printed to a file for specified locations. Initial water content values are not required when the initial stress period is specified as steady state. Thus, initial water-content values must not be specified within the UZF1 input file when the first stress period is steady state. Another option for developing initial water content profiles is to run the model for several years while repeating a cycle of infiltration rates. This step may be carried out subsequent to a steady-state simulation._

I assume the budget table for each uzf reach shown above is for a steady state simulation.  If so, this is why the infiltration exactly equals the GWF term (which is recharge to the water table) and why storage is zero.  I would expect this to look differently for a transient run.

Sorry if I've misunderstood your situation.

By the way, @emorway-usgs is the probably the best person to help with this.  I see he's been responding, but he is also on vacation, so we should keep that in mind.
